common/src/main/java/io/netty/util/TrackedIllegalReferenceCountException.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util;
+
+import io.netty.util.internal.StringUtil;
+
+/**
+ * A special {@link IllegalReferenceCountException} with the ability to track access records
+ */
+public class TrackedIllegalReferenceCountException extends IllegalReferenceCountException {
+
+    private static final long serialVersionUID = 1374377399979428484L;
+
+    public TrackedIllegalReferenceCountException(String message, String accessRecords) {
+        super(message + StringUtil.NEWLINE + accessRecords);
```

Comment:
Since `Record` actually a throwable. Should we let Record and `TrackedIllegalReferenceCountException` constitute a parent-child relationship? like:
```
TrackedIllegalReferenceCountException.cause = head
head.cause = next
...
next.cause = tail
```

Reply:
I am not sure I can follow here. Can you explain what exactly this would buy us ?

Reply:
Of course.
I made a simple test to illustrate:

```
public static void main(String[] args) {

    // here is three record.
    //
    // head.next => second.
    // second.next => tail.
    Record tail = new Record();
    Record second = getBytes(tail);
    Record head = getInt(second);

    head.initCause(second);
    second.initCause(tail);

    NullPointerException t = new NullPointerException();

    t.initCause(head);

    throw t;
}

// mock two record.

private static Record getBytes(Record next) {
    return new Record();
}

private static Record getInt(Record next) {
    return new Record();
}
```

Running the above example, we will get exception information:

```
Exception in thread "main" java.lang.NullPointerException
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:676)
Caused by: 
	at io.netty.util.ResourceLeakDetector.getInt(ResourceLeakDetector.java:690)
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:671)
Caused by: 
	at io.netty.util.ResourceLeakDetector.getBytes(ResourceLeakDetector.java:686)
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:670)
Caused by: 
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:669)
```

If I remove `t.initCause (head)`, it only produces:

```
Exception in thread "main" java.lang.NullPointerException
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:676)
```

The `NullPointerException` in the example is our `TrackedIllegalReferenceCountException`.

Note that `Record` itself is a `Throwable`. So we can use `initCause` to associate `Record` with `TrackedIllegalReferenceCountException`, so that for users, they can directly obtain `Record` information from the stack.

Reply:
I continue to enhance this example, I make a small modification in the Record # toString method:

![image](https://user-images.githubusercontent.com/12047600/71515698-b0f6be00-28df-11ea-8bba-02c6202c5cd2.png)

Running the above example also, I will get:

```
Exception in thread "main" java.lang.NullPointerException
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:679)
Caused by: record info
	at io.netty.util.ResourceLeakDetector.getInt(ResourceLeakDetector.java:693)
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:674)
Caused by: record info
	at io.netty.util.ResourceLeakDetector.getBytes(ResourceLeakDetector.java:689)
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:673)
Caused by: record info
	at io.netty.util.ResourceLeakDetector.main(ResourceLeakDetector.java:672)
```

Reply:
Since `TrackedIllegalReferenceCountException` is an exception, and many times we will not catch it, so when this exception is thrown, it may be a good way to intuitively display the `Record` information displayed in the stack.